source/Missions/Services/Agents/Packets/MovementHandler.cs

```diff
@@ -93,22 +114,68 @@ Mission CurrentMission
             }
         }
 
-        private async void PollAgents()
+        private void Handle_ActionDataChanged(MessagePayload<ActionDataChanged> payload)
         {
-            while (m_AgentPollingCancelToken.IsCancellationRequested == false &&
-                   CurrentMission != null)
+            var delta = GetDelta(payload.What);
+
+            delta.CalculateMovement(payload.What);
+        }
+
+        private void Handle_LookDirectionChanged(MessagePayload<LookDirectionChanged> payload)
+        {
+            var delta = GetDelta(payload.What);
+
+            delta.CalculateMovement(payload.What);
+        }    
+
+        private void Handle_MountDataChanged(MessagePayload<MountDataChanged> payload)
+        {
+            var delta = GetDelta(payload.What);
+
+            delta.CalculateMovement(payload.What);
+        }
+
+        private void Handle_MovementInputVectorChanged(MessagePayload<MovementInputVectorChanged> payload)
+        {
+            var delta = GetDelta(payload.What);
+
+            delta.CalculateMovement(payload.What);
+        }
+
+        private AgentMovementDelta GetDelta(IMovementEvent payload)
+        {
+            if (!_agentRegistry.AgentToId.TryGetValue(payload.Agent, out var payloadGuid))
             {
-                foreach (Guid guid in _agentRegistry.ControlledAgents.Keys)
-                {
-                    Agent agent = _agentRegistry.ControlledAgents[guid];
-                    if (agent.Mission != null)
-                    {
-                        MovementPacket packet = new MovementPacket(guid, agent);
-                        _client.SendAll(packet);
-                    }
-                }
+                throw new ArgumentException($"No {nameof(Agent)} found");
```

Comment:
Have this be logged rather than an exception
```suggestion
                Logger.Error("No {agent} found", nameof(Agent));
                return;
```